### PR TITLE
Problem: BigchainDB and Tendermint inconsistencies because disjoint services

### LINKED ISSUE #2232: Problem: BigchainDB 2.0 crashes on the test network.
## Description

Tendermint version: `0.12.0`
BigchainDB version: `2.0.0a2`

On the BigchainDB test network i.e.
- test.bigchaindb.com
- test2.bigchaindb.com
- test3.bigchaindb.com
- test4.bigchaindb.com

The BigchainDB process crashes which results in the unavailability of BigchainDB i.e. the HTTP API is accessible but  it is not functional.

The reason for that is, when BigchainDB crashes the connection between Tendermint and the proxy app(BigchainDB) is broken. Kubernetes restarts the crashed BigchainDB instance but Tendermint never initiates the connection towards the proxy app. Which is an expected behavior of Tendermint as discussed [here](https://github.com/tendermint/tendermint/issues/243#issuecomment-339327731)

## Solution
- The solution to this problem is two fold:
  - Since BigchainDB and Tendermint are so tightly couple we need to introduce a process supervisor to make the act like a single microservice, so that if BigchainDB crashes, Tendermint does as well and both are restarted and Tendermint requests a connection towards the proxy app.
    - In Kubernetes, they can be exposed as part of a one POD.
    - For BigchainDB as a system service/process, we need to introduce a process supervisor such as `systemd`.
  - The second resolution is to figure out why BigchainDB is crashing. I have provided some logs that might provide some insight.
```
2018-04-23T08:16:03.185023163Z [2018-04-23 08:16:03] [DEBUG] (bigchaindb.tendermint.lib) {'jsonrpc': '2.0', 'id': '5b31d30e-29b5-41ef-82c9-7724b373a31d', 'error': {'code': -32603, 'message': 'Internal error', 'data': 'Error broadcasting transaction: read tcp 10.244.3.70:53982->10.244.2.64:46658: read: connection reset by peer'}} (webapi     - pid: 22)\n
2018-04-23T08:15:26.547564024Z [2018-04-23 08:15:26] [DEBUG] (gunicorn.error) GET / (webapi     - pid: 22)\n
2018-04-23T08:15:30.762913195Z [2018-04-23 08:15:30] [DEBUG] (gunicorn.error) GET / (webapi     - pid: 22)\n
2018-04-23T08:15:31.54679635Z [2018-04-23 08:15:31] [DEBUG] (gunicorn.error) GET / (webapi     - pid: 22)\n
2018-04-23T08:16:16.891529836Z [2018-04-23 08:16:16] [DEBUG] (gunicorn.error) POST /api/v1/transactions (webapi     - pid: 22)\n
2018-04-23T08:16:16.982143203Z [2018-04-23 08:16:16] [DEBUG] (urllib3.connectionpool) Starting new HTTP connection (1): tm-instance-1 (webapi     - pid: 22)\n
2018-04-23T08:16:27.839106752Z \e[32mINFO    \e[0m \e[37mShutting down server\e[0m\n
2018-04-23T08:16:27.839868967Z \e[32mINFO    \e[0m \e[37mShutting down server\e[0m\n
2018-04-23T08:16:27.839890268Z [2018-04-23 08:16:27] [INFO] (abci.app) Shutting down server (MainProcess - pid: 1)\n
2018-04-23T08:16:03.185041164Z [2018-04-23 08:16:03] [ERROR] (bigchaindb.web.views.base) HTTP API error: 500 - POST:/api/v1/transactions - Internal error (webapi     - pid: 22)\n
2018-04-23T08:16:11.37999765Z [2018-04-23 08:16:11] [WARNING] (bigchaindb.tendermint.event_stream) WebSocket connection failed with exception the JSON object must be str, bytes or bytearray, not 'EofStream' (ws_to_tendermint - pid: 17)\n
2018-04-23T08:16:11.54641188Z [2018-04-23 08:16:11] [DEBUG] (gunicorn.error) GET / (webapi     - pid: 22)\n
2018-04-23T08:16:11.546933591Z [2018-04-23 08:16:11] [DEBUG] (gunicorn.error) GET / (webapi     - pid: 22)\n
```

I collected logs from two separate instances and these exit logs seem to be a commonality between them. I can provide more logs if needed.

I haven't really found particular steps to reproduce, so that is slightly unfortunate.

### PR DESCRIPTION
**Pull request by muawiakh**
## Problem
The BigchainDB process crashes which results in the unavailability of BigchainDB i.e. the HTTP API is accessible but it is not functional.

The reason for that is, when BigchainDB crashes the connection between Tendermint and the proxy app(BigchainDB) is broken. Kubernetes restarts the crashed BigchainDB instance but Tendermint never initiates the connection towards the proxy app. Which is an expected behavior of Tendermint as discussed here

## Solution

Since BigchainDB and Tendermint are so tightly coupled we need to introduce a process supervisor to make them act like a single microservice, so that if BigchainDB crashes, Tendermint does as well and both are restarted and Tendermint requests a connection towards the proxy app.
- In Kubernetes, they can be exposed as part of a one POD.
- For BigchainDB as a system service/process, we need to introduce a process supervisor such as systemd.

This PR only solves the former.

## Changes
- Upgrade deployment from Tendermint `v0.12.0` to `v0.19.0`
- Update some documentation
- Fix `nginx-http` entrypoint issues.
- Update `generate-configs.sh` script to handle config generation without `https-certificates`.
- Update Dockerfile to process dependency links introduced by `abci`
- Integrate BigchainDB and Tendermint as a single microservice.
  - This required making BigchainDB to be exposed as a statefulset.
  - Introduce new liveness probe checks.

## Issues Resolved
Partially fixes #2232

### CODE REVIEW

File: k8s/scripts/vars [hunk 1]
```diff
@@ -1,44 +1,43 @@
...
 
 # Comma separated list of initial peers in the
 # network.
-TM_SEEDS='tm-instance-0,tm-instance-1,tm-instance-2'
+BDB_SEEDS='bdb-instance-0'
 
 # Comma separated list of validators in the
 # network
-TM_VALIDATORS='tm-instance-0,tm-instance-1,tm-instance-2'
+BDB_VALIDATORS='bdb-instance-0'
```
Comment (kansi) [inline k8s/scripts/vars]: @muawiakh How do the nodes connect if they don't share the validators?

File: k8s/bigchaindb/tendermint_container/tendermint_entrypoint.bash [hunk 2]
```diff
@@ -48,21 +48,27 @@ else
   echo tm_instance_name="$TM_INSTANCE_NAME"
 fi
 
+echo "wtf"
```
Comment (shahbazn) [inline k8s/bigchaindb/tendermint_container/tendermint_entrypoint.bash]: remove this perhaps :)

File: k8s/scripts/vars [hunk 3]
```diff
@@ -1,44 +1,43 @@
...
-# Tendermint instance name of the bigchaindb
-# node. This name should be unique
-TM_INSTANCE_NAME='tm-instance-0'
+# BigchainDB instance name. This name should be unique
+BDB_INSTANCE_NAME='bdb-instance-0'
 
 # Comma separated list of initial peers in the
 # network.
-TM_SEEDS='tm-instance-0,tm-instance-1,tm-instance-2'
+BDB_SEEDS='bdb-instance-0'
```
Comment (shahbazn) [inline k8s/scripts/vars]: can we please add these back for network configuration

File: k8s/scripts/vars (hunk 1 above)
Comment (shahbazn) [inline k8s/scripts/vars]: same

File: k8s/scripts/vars [hunk 4]
```diff
@@ -1,44 +1,43 @@
...
-TM_VALIDATORS='tm-instance-0,tm-instance-1,tm-instance-2'
+BDB_VALIDATORS='bdb-instance-0'
 
 # Comma separated list of voting
 # power of all validators. Make sure
 # order and number of powers corresponds
-# to TM_VALIDATORS
-TM_VALIDATOR_POWERS='10,10,10'
+# to BDB_VALIDATORS
+BDB_VALIDATOR_POWERS='10'
```
Comment (shahbazn) [inline k8s/scripts/vars]: same

Comment (muawiakh) [reply k8s/scripts/vars]: done.

File: k8s/scripts/vars (hunk 1 above)
Comment (muawiakh) [reply k8s/scripts/vars]: done.

File: k8s/scripts/vars (hunk 3 above)
Comment (muawiakh) [reply k8s/scripts/vars]: done

File: k8s/bigchaindb/tendermint_container/tendermint_entrypoint.bash (hunk 2 above)
Comment (muawiakh) [reply k8s/bigchaindb/tendermint_container/tendermint_entrypoint.bash]: done :)

File: k8s/scripts/vars (hunk 1 above)
Comment (muawiakh) [reply k8s/scripts/vars]: I updated these parameters by mistake, addressed now.

File: k8s/bigchaindb/bigchaindb-ss.yaml [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        imagePullPolicy: Always
+        image: bigchaindb/nginx_pub_key_access:2.0.0-alpha3
+        env:
+        - name: TM_PUB_KEY_ACCESS_PORT
+          valueFrom:
+            configMapKeyRef:
+              name: tendermint-config
+              key: bdb-pub-key-access
+        ports:
+        - containerPort: 9986
```
Comment (shahbazn) [inline k8s/bigchaindb/bigchaindb-ss.yaml]: `bigchaindb-ext-conn-svc.yaml` above exposes `46657` for pub key exchange?

**Pull request merged**